add SVM sensitivity example (#43)
jump-dev/DiffOpt.jl: docs/src/sensitivity-analysis-svm.md

```math
\begin{split}
\begin{array} {ll}
\mbox{minimize} & \sum_{i=1}^{N} \xi_{i} \\
\mbox{s.t.} & \xi_{i} \ge 0 \quad i=1..N  \\
            & y_{i} (w^T X_{i} + b) \ge 1 - \xi[i]\\
\end{array}
\end{split}
```
where
- `X`, `y` are the `N` data points
- `ξ` is the soft-margin loss.

## Define and solve the SVM

Import the libraries.

```@example 1
import Random
using Test
import SCS
import Plots
using DiffOpt
using LinearAlgebra
import MathOptInterface
const MOI = MathOptInterface
nothing # hide
```

Construct separatable, non-trivial data points.
```@example 1
N = 100
D = 2
Random.seed!(6)
X = vcat(randn(N, D), randn(N,D) .+ [4.0,1.5]')
y = append!(ones(N), -ones(N))
nothing # hide
```

Let's define the variables.
```@example 1
model = diff_optimizer(SCS.Optimizer) 
MOI.set(model, MOI.Silent(), true)

# add variables
l = MOI.add_variables(model, N)
w = MOI.add_variables(model, D)
b = MOI.add_variable(model)
nothing # hide
```

Add the constraints.
```@example 1
MOI.add_constraint(
    model,

>     MOI.VectorAffineFunction(
>         MOI.VectorAffineTerm.(1:N, MOI.ScalarAffineTerm.(1.0, l)), zeros(N)
>     ), 
>     MOI.Nonnegatives(N)

Reviewer: ```suggestion
    MOI.VectorAffineFunction(
        MOI.VectorAffineTerm.(1:N, MOI.ScalarAffineTerm.(1.0, l)), zeros(N),
    ), 
    MOI.Nonnegatives(N),
```

end
terms = MOI.VectorAffineTerm.(1:N, Ax)
f = MOI.VectorAffineFunction(
    vec(terms),
    -ones(N),
)
MOI.add_constraint(
    model,
    f,
    MOI.Nonnegatives(N),
)
nothing # hide
```

Define the linear objective function and solve the SVM model.
```@example 1
objective_function = MOI.ScalarAffineFunction(
                        MOI.ScalarAffineTerm.(ones(N), l),
                        0.0,
                    )
MOI.set(model, MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}(), objective_function)
MOI.set(model, MOI.ObjectiveSense(), MOI.MIN_SENSE)

MOI.optimize!(model)

loss = MOI.get(model, MOI.ObjectiveValue())
wv = MOI.get(model, MOI.VariablePrimal(), w)
bv = MOI.get(model, MOI.VariablePrimal(), b)
nothing # hide
```

We can visualize the separating hyperplane. 

```@example 1
# build SVM points
svm_x = [0.0, 3.0]
svm_y = (-bv .- wv[1] * svm_x )/wv[2]

p = Plots.scatter(X[:,1], X[:,2], color = [yi > 0 ? :red : :blue for yi in y], label = "")
Plots.yaxis!(p, (-2, 4.5))
Plots.plot!(p, svm_x, svm_y, label = "loss = $(round(loss, digits=2))", width=3)
Plots.savefig("svm_separating.svg")
nothing # hide
```

![svg](svm_separating.svg)

# Experiments

> Now that we've solved the SVM, we can compute the sensitivity of program variables -- the separating hyperplane in our case -- with respect to perturbations in problem data -- the data points -- using DiffOpt. For illustration, we've explored two questions:

Reviewer: ```suggestion
Now that we've solved the SVM, we can compute the sensitivity of optimal values -- the separating hyperplane in our case -- with respect to perturbations of the problem data -- the data points -- using DiffOpt. For illustration, we've explored two questions:
```

> - How does a change in labels of the data points (`y=1` to `y=-1`, and vice versa) affects the position of the hyperplane? This is achieved by finding gradient of `w`, `b` with respect to `y[i]`, classification label of the ith data point (do not confuse `y[i]` with the array of dual variables.)

Reviewer: ```suggestion
- How does a change in labels of the data points (`y=1` to `y=-1`, and vice versa) affect the position of the hyperplane? This is achieved by finding the gradient of `w`, `b` with respect to `y[i]`, the classification label of the ith data point.
```

Reviewer: This one is tricky to justify with DiffOpt: there is no notion of an infinitesimal perturbation for the labels, I would be for removing it

Reviewer 2: yep. i didn't have any clue about perturbation in categorical data, but the change in value of the label (although not smooth) affects the hyperplane position

- How does a change in coordinates of the data points, `X`, affects the position of the hyperplane? This is achieved by finding gradient of `w`, `b` with respect to `X[i]`, 2D coordinates of the data points.

> Note that the SVM program can be modelled as a conic optimization program

Reviewer: ```suggestion
Note that the SVM program can be modelled as a conic optimization program:
```

> Note that the SVM program can be modelled as a conic optimization program:

Reviewer: ```suggestion
Note that finding the optimal SVM can be modelled as a conic optimization problem:
```

    0 &    0 & ... &    0 & -y_2 X_{2,1} & ... & -y_1 X_{2,N} & -y_2  \\ 
    : &    : & ... &    : &           :  & ... &          :   & :   \\ 
    0 &    0 & ... &    0 & -y_N X_{N,1} & ... & -y_N X_{N,N} & -y_N  \\ 
\end{bmatrix} \\\\

b &= [0, 0, ... 0 \text{(N times)}, l_1 - 1, l_2 -1, ... l_N -1] \\\\

\mathcal{K} &= \text{Set of Nonnegative cones}
\end{align*}
```

## Experiment 1: Gradient of hyperplane wrt the data point labels

Construct perturbations in data point labels `y` without changing the data point coordinates `X`.

```@example 1
ðA = zeros(2*N, N+D+1)
ðb = zeros(2*N)
ðc = zeros(N+D+1)

∇ = Float64[]

# begin differentiating
for Xi in 1:N
    # note that 
    ðA[N+Xi, N+D+1] = 1.0
    
    dx, dy, ds = backward(model, ðA, ðb, ðc)
    dl, dw, db = dx[1:N], dx[N+1:N+1+D], dx[N+1+D]
    push!(∇, norm(dw)+norm(db))
    
    ðA[N+Xi, N+D+1] = 0.0  # reset the change made above
end
LinearAlgebra.normalize!(∇)
nothing # hide
```

Visualize point sensitivities with respect to separating hyperplane. Note that the gradients are normalized.
```@example 1
p2 = Plots.scatter(
    X[:,1], X[:,2], 
    color = [yi > 0 ? :red : :blue for yi in y], label = "",

>     markersize = ∇ * 20

Reviewer: ```suggestion
    markersize = ∇ * 20,
```

)
Plots.yaxis!(p2, (-2, 4.5))
Plots.plot!(p2, svm_x, svm_y, label = "loss = $(round(loss, digits=2))", width=3)
Plots.savefig("sensitivity2.svg")
nothing # hide
```

![](sensitivity2.svg)

## Experiment 2: Gradient of hyperplane wrt the data point coordinates

Similar to previous example, construct perturbations in data points coordinates `X`.
```julia
# constructing perturbations
ðA = zeros(2*N, N+D+1)
ðb = zeros(2*N)
ðc = zeros(N+D+1); 

∇ = Float64[]

# begin differentiating
for Xi in 1:N
    ðA[N+Xi, N.+(1:D+1)] = ones(3)
    
    dx, dy, ds = backward(model, ðA, ðb, ðc)
    dl, dw, db = dx[1:N], dx[N+1:N+1+D], dx[N+1+D]
    push!(∇, norm(dw)+norm(db))
    
    ðA[N+Xi, N.+(1:D+1)] = zeros(3)
end
LinearAlgebra.normalize!(∇)
```

We can visualize point sensitvity with respect to the separating hyperplane. Note that the gradients are normalized.
```@example 1
p3 = Plots.scatter(
    X[:,1], X[:,2], 
    color = [yi > 0 ? :red : :blue for yi in y], label = "",

>     markersize = ∇ * 20

Reviewer: ```suggestion
    markersize = ∇ * 20,
```
